kore/src/main/scala/org/kframework/attributes/Att.scala

```diff
@@ -40,130 +62,245 @@ class Att private (val att: Map[(String, String), Any]) extends AttributesToStri
 }
 
+
 object Att {
+  // The Key constructor is private, enforcing that attributes can only be defined within this object
+  case class Key private[Att](private[Att] val key: String) extends Serializable {
+    override def toString: String = this.key
+    // Att.GOOD_KEY.copy(key="bad-att") would be the same as constructing Att.Key("bad-att"),
+    // so we manually declare copy() to make it private
+    private[Key] def copy(): Unit = ()
+  }
+  object Key {
+    private[Att] def apply(key: String): Key = new Key(key)
+  }
+
+  /*
+   * WARNING: Only use this in exceptional circumstances!
+   */
+  def unsafeRawAttKey(key: String): Att.Key = Att.Key(key)
 
   val empty: Att = Att(Map.empty)
 
-  val ALIAS = "alias"
-  val ALIAS_REC = "alias-rec"
-  val ALL_PATH = "all-path"
-  val ANYWHERE = "anywhere"
-  val ASSOC = "assoc"
-  val BAG = "bag"
-  val BITWIDTH = "bitwidth"
-  val BRACKET = "bracket"
-  val CELL = "cell"
-  val CELL_FRAGMENT = "cellFragment"
-  val CELL_OPT_ABSENT = "cellOptAbsent"
-  val COMM = "comm"
-  val CONCRETE = "concrete"
-  val COOL = "cool"
-  val DIGEST = "digest"
-  val EXPONENT = "exponent"
-  val FUNCTION = "function"
-  val FUNCTIONAL = "functional"
-  val GENERATED_BY_LIST_SUBSORTING = "generatedByListSubsorting"
-  val HEAT = "heat"
-  val HOOK = "hook"
-  val IDEM = "idem"
-  val IMPURE = "impure"
-  val KORE = "kore"
-  val LABEL = "label"
-  val LEMMA = "lemma"
-  val LOCATION = "org.kframework.attributes.Location"
-  val MACRO = "macro"
-  val MACRO_REC = "macro-rec"
-  val MAINCELL = "maincell"
-  val MATCHING = "matching"
-  val NON_EXECUTABLE = "non-executable"
-  val ONE_PATH = "one-path"
-  val ORIGINAL_PRD = "originalPrd"
-  val OWISE = "owise"
-  val PATTERN = "pattern"
-  val PATTERN_FOLDING = "pattern-folding"
-  val PREDICATE = "predicate"
-  val PRIORITY = "priority"
-  val PRIVATE = "private"
-  val PROJ = "proj"
-  val PUBLIC = "public"
-  val RECORD_PRD = "recordPrd"
-  val REFERS_RESTORE_CONFIGURATION = "refers_RESTORE_CONFIGURATION"
-  val REFERS_THIS_CONFIGURATION = "refers_THIS_CONFIGURATION";
-  val SEQSTRICT = "seqstrict"
-  val SIGNIFICAND = "significand"
-  val SIMPLIFICATION = "simplification"
-  val SMTLIB = "smtlib"
-  val SMT_HOOK = "smt-hook"
-  val SMT_LEMMA = "smt-lemma"
-  val SMT_PRELUDE = "smt-prelude"
-  val SOURCE = "org.kframework.attributes.Source"
-  val SPECIFICATION = "specification"
-  val STRICT = "strict"
-  val SYMBOLIC = "symbolic"
-  val SYNTAX_MODULE = "syntaxModule"
-  val TAG = "tag"
-  val TOKEN = "token"
-  val TOP_RULE = "topRule"
-  val TOTAL = "total"
-  val TRANSITION = "transition"
-  val TRUSTED = "trusted"
-  val UNBOUND_VARIABLES = "unboundVariables"
-  val UNIT = "unit"
-  val UNIQUE_ID = "UNIQUE_ID"
-  val USER_LIST = "userList"
+  // Note: Any field Att.UPPER_CASE_NAME of Att.Key type is automatically added to the
+  // whitelist of attribute keys (through reflection).
```

Comment:
See my comment below. Here is the list referenced below of attributes that SHOULD be part of the whitelist:

* alias
* alias-rec
* all-path
* applyPriority
* assoc
* avoid
* bag
* binder
* bracket
* cell
* cellCollecction
* cellName
* color
* colors
* comm
* concrete
* constructor
* context
* cool
* element
* exit
* format
* freshGenerator
* function
* functional
* group
* heat
* hybrid
* hook
* idem
* impure
* initial
* initializer
* injective
* internal
* kast
* klabel
* kore
* label
* latex
* left
* lemma
* locations
* macro
* macro-rec
* maincell
* memo
* mlBinder
* mlOp
* multiplicity
* noThread
* non-assoc
* non-executable
* not-lr1
* one-path
* owise
* parser
* prec
* prefer
* priority
* private
* public
* result
* returnsUnit
* right
* seqstrict
* simplification
* smtlib
* smt-hook
* smt-lemma
* stream
* strict
* symbol
* symbolic
* tag
* token
* topcell
* total
* trusted
* type
* unblock
* unboundVariables
* unit
* unparseAvoid
* unused
* wrapElement

Reply:
Of this list, the following attributes exist on the list because they're used in practice currently, but we should probably evaluate in a follow-up to determine it they still serve a purpose or if they can be removed entirely:

* bag
* noThread
* topcell
* unblock

The following attributes are definitely used in practice in the K prelude, but we may wish to restrict their usage in user written code:

* internal
* mlBinder
* mlOp
* stream
* unparseAvoid

Reply:
Checking all usages of `Att.MY_ATT` and searching all the source files with
``` 
grep -rl 'my-att' . --include \*.hpp --include \*.h --include \*.cpp --include \*.java --include \*.scala --include \*.jj --include \*.jjt --include \*.hs --include \*.x --include \*.y --include \*.l --include \*.k --include \*.md
```
from what I can tell,
- `bag` is both read and written, and seems necessary on a quick glance
- `noThread` appears in some K source files, is added programmatically to productions in the frontend, but is never actually consumed anywhere
- `topcell` never appears in any K source file, is added programmatically, but is never actually consumed anywhere.
- `unblock` is currently only used to identify the rule `stdinUnblock`, so we could just search for the `klabel` instead. (However, there is a TODO comment indicating that we should add more `unblock` rules in the future, which would require the attribute.)

Reply:
Some context:

The bag attribute is automatically added to certain cell collections in a configuration declaration in order to inform the difference between a map and a bag, both of which are non-idempotent, associative, commutative collections. We can probably just use the hook attribute on the sort declaration instead.

noThread is added to certain sentences generated from a configuration declaration, so if we parsed a definition, emitted it back to k, and parsed it again, it would crash if we reject this attribute. But it was created to support a feature of the ocaml backend which has not existed for a considerable amount of time and can probably easily be removed entirely.

topcell is similar to nothread except that it was never used anywhere. Originally it was planned to have meaning when the compiler pipeline was rewritten entirely in the early days of K 4.0, but when it came time to implement that part of the compiler, it was never actually used. It should also be easy to delete.

unblock can probably be replaced with an instance of the label attribute. That todo is old enough and pertains to code that has not been actively used in production in many years, so it can just be ignored.